File: backend/app/api/v1/endpoints/anomalies.py
"""Anomaly detection alerts, event drilldown, and explainability endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query

from backend.app.api.v1.deps import get_repository
from backend.app.core.database import DatabaseRepository
from backend.app.models.processing import AnomalyEventRecord, PaginatedResponse, PaginationMeta
from ml.explainability.schema import ExplanationSummary

router = APIRouter(prefix="/anomalies", tags=["Anomalies & Alerts"])


@router.get("", response_model=PaginatedResponse[AnomalyEventRecord])
async def list_anomalies(
    station_id: Optional[str] = Query(None, description="Filter by station ID"),
    decision: Optional[str] = Query(None, description="Filter by hybrid decision (e.g. PROBABLE_SENSOR_ANOMALY)"),
    severity: Optional[str] = Query(None, description="Filter by severity (e.g. HIGH, CRITICAL)"),
    start_time: Optional[datetime] = Query(None, description="Start timestamp filter"),
    end_time: Optional[datetime] = Query(None, description="End timestamp filter"),
    limit: int = Query(50, ge=1, le=500, description="Items per page"),
    offset: int = Query(0, ge=0, description="Page offset"),
    repo: DatabaseRepository = Depends(get_repository),
):
    """List detected anomalies with multi-criteria filtering and pagination."""
    items, total = repo.get_anomalies(
        station_id=station_id,
        decision=decision,
        severity=severity,
        start_time=start_time,
        end_time=end_time,
        limit=limit,
        offset=offset,
    )
    return PaginatedResponse(
        items=items,
        pagination=PaginationMeta(
            total_count=total,
            limit=limit,
            offset=offset,
            has_more=(offset + limit) < total,
        ),
    )


@router.get("/{event_id}", response_model=AnomalyEventRecord)
async def get_anomaly_detail(
    event_id: str,
    repo: DatabaseRepository = Depends(get_repository),
):
    """Get single anomaly event record by its unique event ID."""
    ev = repo.get_anomaly_by_id(event_id)
    if not ev:
        raise HTTPException(status_code=404, detail=f"Anomaly event '{event_id}' not found.")
    return ev


@router.get("/{event_id}/explanation", response_model=ExplanationSummary)
async def get_anomaly_explanation(
    event_id: str,
    repo: DatabaseRepository = Depends(get_repository),
):
    """Get complete Explainability package (SHAP attributions, neighbor comparison, SOP steps)."""
    exp = repo.get_anomaly_explanation(event_id)
    if not exp:
        raise HTTPException(
            status_code=404,
            detail=f"Explanation package for anomaly event '{event_id}' not found.",
        )
    return exp

File: frontend/src/api/anomalies.ts
/**
 * SkyGuard AI — Anomalies & Explainability API Service
 * Authoritative endpoints:
 * - GET /api/v1/anomalies
 * - GET /api/v1/anomalies/{event_id}
 * - GET /api/v1/anomalies/{event_id}/explanation
 */

import { apiClient, USE_MOCK_DATA } from './client';
import { AnomalyEventRecord, ExplanationSummary, PaginatedResponse } from '../types/api';
import { MOCK_ANOMALIES, getMockAnomalyExplanation } from '../mock/mockAnomalies';

export interface AnomalyFilterParams {
  station_id?: string;
  decision?: string;
  severity?: string;
  start_time?: string;
  end_time?: string;
  limit?: number;
  offset?: number;
}

export const anomaliesApi = {
  async getAnomalies(params?: AnomalyFilterParams): Promise<AnomalyEventRecord[]> {
    if (USE_MOCK_DATA) {
      let list = [...MOCK_ANOMALIES];
      if (params?.station_id) {
        list = list.filter((a) => a.station_id === params.station_id);
      }
      if (params?.decision) {
        list = list.filter((a) => a.decision === params.decision);
      }
      if (params?.severity) {
        list = list.filter((a) => a.severity === params.severity);
      }
      return list;
    }

    const query = new URLSearchParams();
    if (params?.station_id) query.set('station_id', params.station_id);
    if (params?.decision) query.set('decision', params.decision);
    if (params?.severity) query.set('severity', params.severity);
    if (params?.start_time) query.set('start_time', params.start_time);
    if (params?.end_time) query.set('end_time', params.end_time);
    if (params?.limit) query.set('limit', String(params.limit));
    if (params?.offset) query.set('offset', String(params.offset));

    const res = await apiClient.get<PaginatedResponse<AnomalyEventRecord>>(
      `/anomalies?${query.toString()}`
    );
    return res.items;
  },

  async getAnomaly(eventId: string): Promise<AnomalyEventRecord> {
    if (USE_MOCK_DATA) {
      const anom = MOCK_ANOMALIES.find((a) => a.event_id === eventId);
      if (!anom) throw new Error(`Anomaly event ${eventId} not found`);
      return { ...anom };
    }
    return apiClient.get<AnomalyEventRecord>(`/anomalies/${eventId}`);
  },

  async getAnomalyExplanation(eventId: string): Promise<ExplanationSummary> {
    if (USE_MOCK_DATA) {
      return getMockAnomalyExplanation(eventId);
    }
    return apiClient.get<ExplanationSummary>(`/anomalies/${eventId}/explanation`);
  },
};
